### PR TITLE
 Deploy OpenWhisk API Gateway

### PR DESCRIPTION
**Pull request by kameshsampath**
OpenShift resources to deploy [OpenWhisk API Gateway](https://github.com/apache/incubator-openwhisk-apigateway) to Openshift

### CODE REVIEW

**Review by jcrossley3 — CHANGES_REQUESTED** [review]
Looks good, thanks! Would you mind renaming the files to `*.yml` please?

**Comment by kameshsampath** [comment]
@jcrossley3 - renamed the files to `.yml`

File: openshift/apigateway.yml [hunk 1]
```diff
@@ -0,0 +1,137 @@
+########################################
+## API Gateway ConfigMap
+########################################
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: apigateway
+  namespace: openwhisk
```
Comment (bbrowning) [inline openshift/apigateway.yml]: Namespace should be left off here like the other OpenShift configs.

File: openshift/apigateway.yml [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+data:
+  api_host: openwhisk-openwhisk.192.168.64.58.nip.io
```
Comment (bbrowning) [inline openshift/apigateway.yml]: `api_host`, `apigw_url`, and `redis_host` don't need to be hardcoded here. They need to be figured out dynamically in the case of `api-host` and `apigw_url` and perhaps set to just `redis` for redis_host.

File: openshift/apigateway.yml [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+# START REDIS Deployment
+########################################
+---
+apiVersion: v1
+kind: Service
+metadata:
+  labels:
+    app: redis
+  name: redis
+  namespace: openwhisk
```
Comment (bbrowning) [inline openshift/apigateway.yml]: Leave namespace off here as well.

File: openshift/apigateway.yml [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+  selector:
+    app: redis
+---
+apiVersion: apps.openshift.io/v1
+kind: DeploymentConfig
+metadata:
+  labels:
+    app: redis
+  name: redis
+  namespace: openwhisk
```
Comment (bbrowning) [inline openshift/apigateway.yml]: Leave namespace off.

File: openshift/apigateway.yml [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+# START OpenWhisk ApiGateway Deployment
+########################################
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: apigateway
+  labels:
+    name: apigateway
+  namespace: openwhisk
```
Comment (bbrowning) [inline openshift/apigateway.yml]: Leave namespace off.

File: openshift/apigateway.yml [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    - port: 8080
+      name: mgmt
+    - port: 9000
+      name: api   
+---
+apiVersion:  apps.openshift.io/v1
+kind: DeploymentConfig
+metadata:
+  name: apigateway
+  namespace: openwhisk
```
Comment (bbrowning) [inline openshift/apigateway.yml]: Leave namespace off.

File: openshift/apigateway.yml [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+    - type: ConfigChange
+
+---
+apiVersion: route.openshift.io/v1
+kind: Route
+metadata:
+  labels:
+    app: apigateway
+  name: apigateway
+  namespace: openwhisk
```
Comment (bbrowning) [inline openshift/apigateway.yml]: Leave namespace off.

File: openshift/apimangement.yml [hunk 8]
```diff
@@ -0,0 +1,39 @@
+apiVersion: batch/v1
```
Comment (bbrowning) [inline openshift/apimangement.yml]: typo in this filename - should be `apimanagement.yml`

File: openshift/apimangement.yml [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+kind: Job
+metadata:
+  name: install-apimanagement
+  namespace: openwhisk
```
Comment (bbrowning) [inline openshift/apimangement.yml]: Leave namespace off.

**Review by bbrowning — CHANGES_REQUESTED** [review]
Thanks for the PR! I've made some review requests - let me know if you need help figuring this out. Specifically, removing the need to hardcode hostnames and URL may take some thought to figure out.

File: openshift/apigateway.yml (hunk 1 above)
Comment (kameshsampath) [reply openshift/apigateway.yml]: done!

File: openshift/apigateway.yml (hunk 2 above)
Comment (kameshsampath) [reply openshift/apigateway.yml]: @bbrowning  - actually I thought to deduce that, but don't know how we can do that - if we can do that then I don't need to  have those `api_host` and `apigw_url` entries.   Thoughts ?

File: openshift/apigateway.yml (hunk 3 above)
Comment (kameshsampath) [reply openshift/apigateway.yml]: done

File: openshift/apigateway.yml (hunk 5 above)
Comment (kameshsampath) [reply openshift/apigateway.yml]: done

File: openshift/apigateway.yml (hunk 6 above)
Comment (kameshsampath) [reply openshift/apigateway.yml]: done

File: openshift/apimangement.yml (hunk 8 above)
Comment (kameshsampath) [reply openshift/apimangement.yml]: fixed. thanks

File: openshift/apimangement.yml (hunk 9 above)
Comment (kameshsampath) [reply openshift/apimangement.yml]: done

**Comment by kameshsampath** [comment]
@bbrowning - i have fixed all the comments that you had. Please review and merge when you are at it . Currently there will be one parameter `REDIS_HOST` whose values I need to have fully qualified because of minishift bug https://github.com/minishift/minishift/issues/2070, once that bug is fixed we can remove the FQDN of `REDIS_HOST`

**Review by jcrossley3 — APPROVED** [review]

**Comment by kameshsampath** [comment]
Opened new PR https://github.com/projectodd/openwhisk-openshift/pull/4 and will track it there.

**Pull request closed without merging**